src/main/java/com/shapesecurity/salvation2/Policy.java

```diff
@@ -0,0 +1,1126 @@
+					}
+					this.upgradeInsecureRequests = true;
+				} else {
+					wasDupe = true;
+				}
+				newDirective = new Directive(values);
+				break;
+			}
+			default: {
+				FetchDirectiveKind fetchDirectiveKind = FetchDirectiveKind.fromString(lowcaseDirectiveName);
+				if (fetchDirectiveKind != null) {
+					SourceExpressionDirective thisDirective = new SourceExpressionDirective(values, directiveErrorConsumer);
+					if (this.fetchDirectives.containsKey(fetchDirectiveKind)) {
+						wasDupe = true;
+					} else {
+						this.fetchDirectives.put(fetchDirectiveKind, thisDirective);
+					}
+					newDirective = thisDirective;
+					break;
+				}
+				directiveErrorConsumer.add(Severity.Warning, "Unrecognized directive " + lowcaseDirectiveName, -1);
+				newDirective = new Directive(values);
+				break;
+			}
+		}
+
+		this.directives.add(new NamedDirective(name, newDirective));
+		if (wasDupe) {
+			directiveErrorConsumer.add(Severity.Warning, "Duplicate directive " + lowcaseDirectiveName, -1);
+		}
+		return newDirective;
+	}
+
+	// Note that this removes all directives matching this name.
+	// Returns true if at least one directive was removed.
+	public boolean remove(String name) {
+		boolean removed = false;
+		String lowcaseName = name.toLowerCase(Locale.ENGLISH);
+		ArrayList<NamedDirective> copy = new ArrayList<>(this.directives.size());
+		for (NamedDirective existing : this.directives) {
+			if (!existing.lowcaseName.equals(lowcaseName)) {
+				copy.add(existing);
+			} else {
+				removed = true;
+			}
+		}
+		if (!removed) {
+			return false;
+		}
+		this.directives = copy;
+		switch (lowcaseName) {
+			case "base-uri": {
+				this.baseUri = null;
+				break;
+			}
+			case "block-all-mixed-content": {
+				this.blockAllMixedContent = false;
+				break;
+			}
+			case "form-action": {
+				this.formAction = null;
+				break;
+			}
+			case "frame-ancestors": {
+				this.frameAncestors = null;
+				break;
+			}
+			case "navigate-to": {
+				this.navigateTo = null;
+				break;
+			}
+			case "plugin-types": {
+				this.pluginTypes = null;
+				break;
+			}
+			case "report-to": {
+				this.reportTo = null;
+				break;
+			}
+			case "report-uri": {
+				this.reportUri = null;
+				break;
+			}
+			case "sandbox": {
+				this.sandbox = null;
+				break;
+			}
+			case "upgrade-insecure-requests": {
+				this.upgradeInsecureRequests = false;
+				break;
+			}
+			default: {
+				FetchDirectiveKind fetchDirectiveKind = FetchDirectiveKind.fromString(lowcaseName);
+				if (fetchDirectiveKind != null) {
+					this.fetchDirectives.remove(fetchDirectiveKind);
+				}
+				break;
+			}
+		}
+		return true;
+	}
+
+
+	@Override
+	public String toString() {
+		StringBuilder out = new StringBuilder();
+		boolean first = true;
+		for (NamedDirective directive : this.directives) {
+			if (!first) {
+				out.append("; "); // The whitespace is not strictly necessary but is probably valuable
+			}
+			first = false;
+			out.append(directive.name);
```

Comment:
Why not `lowercasedName`? This is the only use of `name`.

Reply:
The intention is to [preserve case](https://github.com/shapesecurity/salvation/blob/13ef06db4e52a219d40440035efe6f4613167df4/src/main/java/com/shapesecurity/salvation2/Policy.java#L41).

Reply:
Why do you want that? I could understand preserving casing if you're also preserving whitespace, order, and empty directives so that a round trip through Salvation doesn't cause unnecessary diffs. But that's not the case, so I don't understand the value.

Reply:
> Why do you want that?  

Because I've seen multiple people in the wild have tests which assert on the case-sensitive presence of directive names in their policies.

> I could understand preserving casing if you're also preserving whitespace, order, and empty directives

We are [preserving order](https://github.com/shapesecurity/salvation/blob/13ef06db4e52a219d40440035efe6f4613167df4/src/main/java/com/shapesecurity/salvation2/Policy.java#L42) (it's literally the next line after the previous link). I didn't think anyone was likely to be doing weird things with whitespace and asserting that those weird things are preserved; certainly I haven't seen that in the wild. But if you think it's likely to come up I can add it.